parser: partially align use of loc_ghost with what the documentation says

v-gb opened the pull request:
`loc.loc_ghost` is supposed to false iff the ast described by `loc` is present at that location. The parser doesn't follow this condition consistently, which can result in unpredictable results in ocamlformat, where loc_ghost is sometimes consulted.

I fixed cases created by the rules for`keyword%extension` kind of syntax, as these seems like the main offenders I've seen in ocamlformat.

But there are more failures found by the check. Some may be harmless (`+foo` creates a `~+` node whose location refers to `+`), some look weirder (empty non-ghost location with `object[@foo]` in some case).

In the changed test, the parser change should imply that a Pext_extension and its child swap their ghost boolean. But in some cases, the diff shows only ghost locations being added, which confused me. The reason is that `new%foo x` has a top node with a ghost location, while `(new%foo x)` creates the same top node with a non-ghost location.

v-gb committed:
parser: partially align use of loc_ghost with what the documentation says

Mostly in:
- the productions for keyword%extension syntaxes
- the implied 'self in `object ('self) end`

v-gb committed:
parser: regenerate parser.ml

v-gb committed:
add an invariant function to check that loc_ghost is set correctly

spoiler alert: it isn't

gasche commented:
This discussion would be easier to follow if you could repeat (in your PR message or here) what the documentation says about ghost nodes -- it is documented in a somewhere obscure place (introduced in b06262074d043474f8a128fcdeafd831af7de300) and I suspect most people are unfamiliar with this proposed specification.

v-gb commented:
> This discussion would be easier to follow if you could repeat (in your PR message or here) what the documentation says about ghost nodes -- it is documented in a somewhere obscure place (introduced in [b062620](https://github.com/ocaml/ocaml/commit/b06262074d043474f8a128fcdeafd831af7de300)) and I suspect most people are unfamiliar with this proposed specification.

I adjusted the pr description.

Octachron commented on driver/pparse.ml:
```diff
@@ -177,6 +177,18 @@ let set_input_lexbuf ic =
   lexbuf
 
+let check_loc_ghost (type a) (kind : a ast_kind) (ast : a) ~inputfile =
+  if Option.is_some (Sys.getenv_opt "OCAML_LOC_GHOST_CHECK") then
```
Having this check enabled on an environment variable does not sound optimal to me.
Moreover, my impression is that this check is mostly useful for the compiler, ocamlformat and ppxlib developers,
rather than most users.
Thus I think it might be better to just expose this check in the `pparse` API and let the compiler testsuite use it directly.

v-gb replied:
The variable is definitely not intended to be user facing.

My thinking has been that the whole test suite should run with this, at least if the perf impact is negligible. This way, any parser change that's tested somewhere would have its ghostness checked (whereas a test dedicated to this invariant would have limited coverage, because no one would update it when updating the parser).

But this invariant doesn't hold in a number of places, so I can't really do that. So instead this is partial progress: we now have an executable specification of the invariant, but it is used only manually via the env var. With more work (which I don't currently plan to do, but may if the need arises), it could be turned on in some test, e.g. testsuite/tests/parsing, as that seems to have a good coverage of syntax.

I don't really know what I would do with a `?check_ghost_locs:bool` argument in the api of pparse.

gasche replied:
Maybe we could have a `-dparsetree-ghost-locs-invariants` (or maybe `-dparsetree-invariants-ghost-locs`) command-line flag on compiler invocation? We already have `-dcmm-invariants` and `-dflambda-invariants` in ocamlopt.

(A cynical way to say this would be: so far we have been careful in trying not to read random environment variables, and we would want to keep some discipline there. On the other hand the command-line options are already a dumpster fire, so we could add one more user-irrelevant thing in this flat namespace.)

Octachron replied:
We could also also have a `Clflags` variable (`ast_linter`?) without a CLI flag. Typically, the test for the AST pretty printer is already using compiler-libs, and it could easily set this new flag.

v-gb replied:
I replaced the env var by `-dparsetree-ghost-loc-invariants`, and added that flag to a test that exercises a lot of the syntax I'm touching. There are still two invariants failures in there (down from a screenful, as you can see in the first commit), but we can just promote the invariant failure, no need to fix absolutely everything first.

Octachron commented on parsing/ast_invariants.ml:
```diff
@@ -216,3 +216,76 @@ let iterator =
+    ; typ =
+        limit_quadratic_complexity (fun s -> s.typ)
+          (fun v -> check (fun s -> s.typ) Parse.core_type v v.ptyp_loc )
+    }
```
Naive question: do you only check expressions, type expressions, and patterns because they are the most relevant or because other constructs produce less (or no) ghost locations?

v-gb replied:
I can't say I remember my decision making, but I presumably only tested the syntactic categories I was modifying. But yes, I expect most problems (in frequency) to be in these categories.
In principle, we could test more, but it doesn't seem helpful to do that without also spending time fixing the resulting errors, otherwise I think the main effect would be more errors to sift through.

Octachron commented on parsing/parser.mly:
```diff
@@ -437,32 +437,40 @@ let wrap_type_annotation ~loc newtypes core_type body =
+let mktyp_attrs ~loc d ext_attrs =
+  wrap_typ_attrs ~loc (if Option.is_some (fst ext_attrs)
+                       then ghtyp ~loc d
+                       else mktyp ~loc d) ext_attrs
```
Since `mktyp_attrs` is the sole user of `wrap_typ_attrs`, I would propose to merge the two functions (same for the `pat_attrs` functions).

v-gb replied:
I did that, and the functions are much clearer, thanks. In fact, I refactored all the functions this way by removing the calls to the inner function as necessary.

Octachron commented on parsing/ast_invariants.ml:
```diff
@@ -216,3 +216,91 @@ let iterator =
+          (fun v ->
+            check
+              (* ~print:(fun f ty -> Printast.payload 0 f (PTyp ty)) *)
+              (fun s -> s.typ) Parse.core_type v v.ptyp_loc )
```
Note for later: the more generic parsing function would be `poly_type` rather than `core_type`.

Octachron commented on parsing/ast_invariants.ml:
```diff
@@ -216,3 +216,91 @@ let iterator =
+    }
+  in
+  (meth self) self v;
+  if false && not !ok then failwith "invariant failed"
```
This debugging line can be removed, isn't it?

Octachron reviewed (approved):
The changes make the syntax keyword%ext follows the specification for ghost locations, and the test looks useful. Thus the PR looks ready to merge once:
- the menhir bootstrap are localized to a single commit
- the last debugging line is removed

v-gb committed:
parser: refactor previous commit, by removing error prone functions

v-gb commented:
Thanks!
I addressed the various things you mentioned. I squashed some of the existing commits, without really changing them (just removed the debugging line, and maybe some other minor thing like that), and I added the refactor you mentioned on top, so you may want to read just that one commit.

Octachron commented on parsing/parser.mly:
```diff
@@ -3780,8 +3765,9 @@ tuple_type:
-      { wrap_typ_attrs ~loc:$sloc (reloc_typ ~loc:$sloc package_type) attrs }
+  | LPAREN MODULE ext_attrs = ext_attributes package_type = package_type RPAREN
+      { mktyp_attrs ~loc:$sloc
+          (reloc_typ ~loc:$sloc package_type).ptyp_desc ext_attrs }
```
Note for later: this is correct because the type from `package_type` doesn't have attributes currently (the attributes are now stored inside the `Package_type` record themselves.

v-gb replied:
This code should be:

```
  | LPAREN MODULE ext_attrs = ext_attributes package_type = package_type_ RPAREN
      { mktyp_attrs ~loc:$sloc package_type ext_attrs }
```

I can make the change, unless you'd rather I leave the pull request as is.

Octachron replied:
You are right, there is only one valid concrete location here, and thus the call to `reloc_typ` is erroneous here. 
If you have the time to make the change, it would be welcome.

v-gb replied:
Done.
